pubsub/iterator.go

```diff
@@ -523,48 +591,94 @@ func (it *messageIterator) sendAck(m map[string]*AckResult) {
 // percentile in order to capture the highest amount of time necessary without
 // considering 1% outliers. If the ModAck RPC fails and exactly once delivery is
 // enabled, we retry it in a separate goroutine for a short duration.
-func (it *messageIterator) sendModAck(m map[string]*AckResult, deadline time.Duration, logOnInvalid bool) {
+func (it *messageIterator) sendModAck(m map[string]*AckResult, deadline time.Duration, logOnInvalid, isReceipt bool) {
 	deadlineSec := int32(deadline / time.Second)
+
+	isNack := deadline == 0
+
 	ackIDs := make([]string, 0, len(m))
-	for k := range m {
-		ackIDs = append(ackIDs, k)
+	for ackID := range m {
+		ackIDs = append(ackIDs, ackID)
+
 	}
 	it.eoMu.RLock()
 	exactlyOnceDelivery := it.enableExactlyOnceDelivery
 	it.eoMu.RUnlock()
 	var toSend []string
 	for len(ackIDs) > 0 {
 		toSend, ackIDs = splitRequestIDs(ackIDs, ackIDBatchSize)
-		if deadline == 0 {
+
+		var eventStart, eventEnd string
+		if isNack {
 			recordStat(it.ctx, NackCount, int64(len(toSend)))
+			eventStart = eventNackStart
+			eventEnd = eventNackEnd
 		} else {
 			recordStat(it.ctx, ModAckCount, int64(len(toSend)))
+			eventStart = eventModackStart
+			eventEnd = eventModackEnd
 		}
-		addModAcks(toSend, deadlineSec)
-		// Use context.Background() as the call's context, not it.ctx. We don't
-		// want to cancel this RPC when the iterator is stopped.
-		cctx, cancel2 := context.WithTimeout(context.Background(), 60*time.Second)
-		defer cancel2()
-		err := it.subc.ModifyAckDeadline(cctx, &pb.ModifyAckDeadlineRequest{
-			Subscription:       it.subName,
-			AckDeadlineSeconds: deadlineSec,
-			AckIds:             toSend,
-		})
-		if exactlyOnceDelivery {
-			resultsByAckID := make(map[string]*AckResult)
+
+		// Use of anonymous local function allows span.End to be deferred to end of each loop.
+		func() {
+			numBatch := len(toSend)
+			links := make([]trace.Link, 0, numBatch)
 			for _, ackID := range toSend {
-				resultsByAckID[ackID] = m[ackID]
+				if it.enableTracing {
+					// get the parent span context for this ackID for otel tracing.
+					s, _ := it.activeSpans.Load(ackID)
```

Comment:
Should this be `LoadAndDelete` when `isNack`? You're only deleting in `handleKeepAlives` but if the message is nack'd then it's deleted from `keepAliveDeadlines` and won't ever be removed from the sync.Map.